### PR TITLE
fix(es/transforms): Fix `this` in async arrow class properties

### LINKED ISSUE #3228: "this" inside arrow in a static class property is compiled to (void 0)
### Describe the bug

Using `this` inside a fat arrow function on a static class property does not work, because this is compiled to `(void 0)`;

### Input code

```typescript
export class StaticTestClass {
    static testProp = 'Hello world!';

    static testMethod = () => {
        console.log(this.testProp);
    }
}
```

### Config

```json
{
  "jsc": {
    "parser": {
      "syntax": "typescript",
      "tsx": true
    },
    "target": "es2015",
    "loose": false,
    "minify": {
      "compress": false,
      "mangle": false
    }
  },
  "module": {
    "type": "commonjs"
  },
  "minify": false,
  "isModule": true,
  "env": {
    "targets": "node 12"
  }
}
```

### Playground link

https://play.swc.rs/?version=1.2.128&code=H4sIAAAAAAAAA0utKMgvKlFIzkksLlYILkksyUwOSS0ucQbzq7kUgKAYLKpQAhQOKMovULBVUPdIzcnJVyjPL8pJUVS35kJX5ptakpGfAlSooalgawc1BgSS8%2FOK83NS9XLy0zVKMjKL9WCGalqD1dRy1QIAgrT4spEAAAA%3D&config=H4sIAAAAAAAAA0VOSwqDMBC9SnlrF1Xoxjv0EEFHieTHTCyVkLt3IhZ3j%2Fcv2GTCWJAMC3FDcoRsvhiRj0QysU0ZHbIolXmnqtjwSlkdJMOzf6nqYhTCuBgn1MHbYJejdU3RJyaRWzJhdX9n1S4f570R5ZzTTo34GDZBvYuusJX35W5HOlD4nMHzjm4gxJke%2FYBaf%2FJ6D0jYAAAA

### Expected behavior

Expected the `testMethod` method to compile to:
```
StaticTestClass.testMethod = ()=>{
    console.log(StaticTestClass.testProp);
};
```

### Actual behavior

The `testMethod` method compiles to:
```
StaticTestClass.testMethod = ()=>{
    console.log((void 0).testProp);
};
```

### Version

1.2.128

### Additional context

Only seems to happen with typescript in combination with commonjs.

### PR DESCRIPTION
**Pull request by kdy1**
swc_ecma_transforms_compat:
 - `async_to_generator`: Fix `this` in async class properties. (https://github.com/vercel/next.js/issues/33283)
 - Add a test for #3228 (Closes #3228).

---

I created a new PR because I thought I messed up something, but found that it's correct change for many cases.

### CODE REVIEW

File: crates/swc_ecma_transforms_compat/tests/es2017_async_to_generator.rs [hunk 1]
```diff
@@ -1707,28 +1707,26 @@ function s(x) {
...
+              function* (y, a) {
                   let r = function () {
                       var _ref = _asyncToGenerator(
-                          (function* (z, b, ...innerArgs) {
+                          function* (z, b, ...innerArgs) {
                               yield z;
                               console.log(this, innerArgs, arguments);
                               return this.x;
-                          }).bind(_this)
-                      ).bind(_this);
```
Comment (kdy1) [inline crates/swc_ecma_transforms_compat/tests/es2017_async_to_generator.rs:1722]: @Austaras @magic-akari Do you have any idea about `.bind(this)`?
`this` in the current context is `this` of anonymous function created by `let r = function () {`

Comment (Austaras) [reply crates/swc_ecma_transforms_compat/tests/es2017_async_to_generator.rs:1722]: I believe if r is an arrow function, bind is still needed(for now)

Comment (Austaras) [reply crates/swc_ecma_transforms_compat/tests/es2017_async_to_generator.rs:1722]: Oh, it's fine then

Comment (kdy1) [reply crates/swc_ecma_transforms_compat/tests/es2017_async_to_generator.rs:1722]: @kwonoj Yes, and I fixed it using `FnEnvState` and `FnEnvHoister`.

Comment (magic-akari) [reply crates/swc_ecma_transforms_compat/tests/es2017_async_to_generator.rs:1722]: It's good to see that this issue has been fixed :)

Comment (kwonoj) [reply crates/swc_ecma_transforms_compat/tests/es2017_async_to_generator.rs:1722]: guess we can close #2716 even?

Comment (kdy1) [reply crates/swc_ecma_transforms_compat/tests/es2017_async_to_generator.rs:1722]: Not yet, https://github.com/swc-project/swc/blob/d03824b13d8d744f2b16698dbed182e3afe88374/crates/swc_ecma_transforms_compat/src/es2017/async_to_generator.rs#L135 is the last one.

**Comment by magic-akari** [comment]
Hi @kdy1 ,
It seems that #3228 is not related to async, but to class static, which is not fixed.

**Comment by kdy1** [comment]
@magic-akari Actually it's already fixed so I just added [a test](https://github.com/kdy1/swc/blob/0b0589225afb3b79f1ee25aeea952d7e738a6c02/crates/swc_ecma_transforms_compat/tests/fixture/async-to-generator/issue-3228/1/exec.js) to ensure that it's fixed.

The title of PR is a bit misleading...

**Comment by kdy1** [comment]
You can copy-paste the output from https://play.swc.rs/?version=1.2.129&code=H4sIAAAAAAAAA0vOSSwuVgguSSzJTA5JLS5xBvOrebkUgKAYLKxQAhQPKMovULBVUPdIzcnJVyjPL8pJUVS35uXCUOibWpKRnwJUqqGpYGsHMwkEkvPzivNzUvVy8tM1SjIyi%2FVg5mpaQxTV8nLVgkxEc40ewlgNTQDmPQglsAAAAA%3D%3D&config=H4sIAAAAAAAAA0WOSwrDMAxE7zJrb9uF79BDCFcJLv4hOVBjfPfaISU7MfPmoY6POtiOQqIs69KWKn1hwS6SOvGlwkxsRlUOHgaVZOe6CH3MKuSsDLtRUDaIPvmtLZHLsQir3hWlPfzJMUUxv48VdNRW%2BBQ%2BMW7HtfP6usDzgR%2Fs3ILttAAAAA%3D%3D and run it using chrome

image

**Comment by wilcokaartje2go** [comment]
@kdy1 I think @magic-akari is right, the static method does not seem to be fixed when using typescript in combination with commonjs.

https://play.swc.rs/?version=1.2.129&code=H4sIAAAAAAAAA0vOSSwuVgguSSzJTA5JLS5xBvOrebkUgKAYLKxQAhQPKMovULBVUPdIzcnJVyjPL8pJUVS35uXCUOibWpKRnwJUqqGpYGsHMwkEkvPzivNzUvVy8tM1SjIyi%2FVg5mpaQxTV8nLVgkxEc40ewlgNTQDmPQglsAAAAA%3D%3D&config=H4sIAAAAAAAAA0WOSwrDMAxE76J1tt3kDj2EcZXg4B8aBWqM7145pGQ3zOdJnQ54WjtVJ2CZCi2r%2B9JK2irDS6hKCynMUjl5mHays1qD8bIolgKmdXMRvFAKOWxtgnxJVRh4Ipf3%2BG8OA6XyOafRr1sGtEkq%2BQCNB3SPA953%2B%2FriB8tcjwO5AAAA

**Comment by kdy1** [comment]
I found that I added a test to the wrong directory. (I made a mistake because I reverted some pass changes.)
Reopening

**Pull request merged**